src/Util.php

```diff
@@ -132,4 +132,171 @@ private static function compareArrays(array $valueOne, array $valueTwo) : bool
+                if ($onlyOne) {
+                    $type = JsonBrowser::TYPE_INTEGER;
+                } else {
+                    $type |= JsonBrowser::TYPE_INTEGER;
+                }
+            }
+            return $type;
+        }
+
+        if (is_array($value)) {
+            return JsonBrowser::TYPE_ARRAY;
+        }
+
+        if (is_object($value)) {
+            return JsonBrowser::TYPE_OBJECT;
+        }
+
+        throw new Exception(JsonBrowser::ERR_UNKNOWN_TYPE, 'Unknown type: %s', gettype($value)); // @codeCoverageIgnore
+    }
+
+    /**
+     * Cast a value to conform to the given type mask, losing as little fidelity as possible
+     *
+     * @since 2.4.0
+     *
+     * @param int $asType  The type mask to cast to
+     * @param mixed $value The value to cast
+     * @return mixed The cast value
+     */
+    public static function cast(int $asType, $value)
+    {
+        // get the value type
+        $type = self::typeMask($value);
+
+        // check whether value is already one of the desired types
+        if ($type & $asType) {
+            return $value;
+        }
+
+        // cast objects & arrays
+        // -> directly to an object or associative array
+        // -> to a json-encoded string
+        // -> to an integer count of the members
+        // -> to a boolean indicating whether any members are present
+        if ($type & (JsonBrowser::TYPE_OBJECT | JsonBrowser::TYPE_ARRAY)) {
+            if ($asType & JsonBrowser::TYPE_OBJECT) {
+                return (object) $value;
+            } elseif ($asType & JsonBrowser::TYPE_ARRAY) {
+                return (array) $value;
+            } elseif ($asType & JsonBrowser::TYPE_STRING) {
+                return json_encode($value, \JSON_PRETTY_PRINT | \JSON_UNESCAPED_SLASHES | \JSON_UNESCAPED_UNICODE);
+            } elseif ($asType & (JsonBrowser::TYPE_NUMBER | JsonBrowser::TYPE_INTEGER)) {
+                return count((array)$value);
+            } elseif ($asType & JsonBrowser::TYPE_BOOLEAN) {
+                return (bool) count((array)$value);
+            }
+        }
+
+        // cast strings
+        // -> to an array of unicode characters
+        // -> to an array of unicode characters, cast as an object (stdClass)
+        // -> to an integer count of [unicode] characters in the string
+        // -> to a boolean indicating whether the string length is greater than zero
+        if ($type & JsonBrowser::TYPE_STRING) {
+            if ($asType & JsonBrowser::TYPE_ARRAY) {
+                return preg_split('//u', $value, null, \PREG_SPLIT_NO_EMPTY);
+            } elseif ($asType & JsonBrowser::TYPE_OBJECT) {
+                return (object) preg_split('//u', $value, null, \PREG_SPLIT_NO_EMPTY);
+            } elseif ($asType & (JsonBrowser::TYPE_NUMBER | JsonBrowser::TYPE_INTEGER)) {
+                return mb_strlen($value);
+            } elseif ($asType & JsonBrowser::TYPE_BOOLEAN) {
+                return (bool) strlen($value);
```

Comment:
Is all this casting stuff going to wind up driving coerce functionality? I can't remember where we wound up with string "true" and "false" coercing...

Reply:
Possibly, but that's not my motivator for adding this. I'm using this library in another project which needs casting that will always succeed.